swap/swap.go

```diff
@@ -1,98 +1,451 @@
+		return nil, err
+	}
+	lastCheque := s.cheques[peer]
+
+	serial := uint64(1)
+	if lastCheque != nil {
+		cheque = &Cheque{
+			ChequeParams: ChequeParams{
+				Serial: lastCheque.Serial + serial,
+				Amount: lastCheque.Amount + amount,
+			},
+		}
+	} else {
+		cheque = &Cheque{
+			ChequeParams: ChequeParams{
+				Serial: serial,
+				Amount: amount,
+			},
+		}
+	}
+	cheque.ChequeParams.Timeout = defaultCashInDelay
+	cheque.ChequeParams.Contract = s.owner.Contract
+	cheque.ChequeParams.Honey = honey
+	cheque.Beneficiary = beneficiary
+
+	cheque.Signature, err = s.signContent(cheque)
+
+	return cheque, err
+}
+
+// Balance returns the balance for a given peer
+func (s *Swap) Balance(peer enode.ID) (int64, error) {
+	var err error
+	// check the balance in memory
+	peerBalance, ok := s.balances[peer]
+	// if not present, check in disk
+	if !ok {
+		err = s.store.Get(balanceKey(peer), &peerBalance)
+	}
+	return peerBalance, err
+}
+
+// Balances returns the balances for all known SWAP peers
+func (s *Swap) Balances() (map[enode.ID]int64, error) {
+	balances := make(map[enode.ID]int64)
+
+	// add in-memory balances
+	for peerID, peerBalance := range s.balances {
+		balances[peerID] = peerBalance
+	}
+
+	// add store balances, if peer was not already added
+	balanceIterFunction := func(key []byte, value []byte) (stop bool, err error) {
+		peerID := keyToID(string(key), balancePrefix)
+		if _, peerHasBalance := balances[peerID]; !peerHasBalance {
+			var peerBalance int64
+			err = json.Unmarshal(value, &peerBalance)
+			if err == nil {
+				balances[peerID] = peerBalance
+			}
+		}
+		return stop, err
+	}
+	err := s.store.Iterate(balancePrefix, balanceIterFunction)
+	if err != nil {
+		return nil, err
+	}
+
+	return balances, nil
+}
+
+// loadLastSentCheque loads the last cheque for a peer from the state store (persisted)
+func (s *Swap) loadLastSentCheque(peer enode.ID) (err error) {
 	//only load if the current instance doesn't already have this peer's
-	//balance in memory
-	if _, ok := s.balances[peerID]; !ok {
-		err = s.stateStore.Get(peerID.String(), &peerBalance)
-		s.balances[peerID] = peerBalance
+	//last cheque in memory
+	var cheque *Cheque
+	if _, ok := s.cheques[peer]; !ok {
+		err = s.store.Get(sentChequeKey(peer), &cheque)
+		if err == nil {
+			s.cheques[peer] = cheque
+		}
 	}
+	return err
+}
+
+// loadLastReceivedCheque loads the last received cheque for peer
+func (s *Swap) loadLastReceivedCheque(peer enode.ID) (cheque *Cheque) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	s.store.Get(receivedChequeKey(peer), &cheque)
 	return
 }
 
-//Clean up Swap
-func (swap *Swap) Close() {
-	swap.stateStore.Close()
+// saveLastReceivedCheque saves cheque as the last received cheque for peer
+func (s *Swap) saveLastReceivedCheque(peer enode.ID, cheque *Cheque) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	return s.store.Put(receivedChequeKey(peer), cheque)
+}
+
+// Close cleans up swap
+func (s *Swap) Close() {
```

Comment:
`Swap` has a `Stop` method (which is masked from the beholder's eye since its in a different file in this package). Please get rid of this function and do the shutdown in the `Stop()` method. I also would be happy to see all of that other stuff in that protocols file in this file. As I already commented - too many files in this package for my taste

Reply:
the `Stop()` function in `protocols` just logs and always returns `nil`. so i'm not sure what we should do here.

Reply:
But if we close the store in `Stop`, we have to call Stop manually in tests, which is not good really. Stop should only be called by the stack.

Reply:
Solved in #1664.

Reply:
my intent was to _remove_ `Close`, and have the logic sit in `Stop`. I don't understand WDYM by 

> we have to call Stop manually in tests

Reply:
@acud `Start` and `Stop` are functions we are using pretty much wrong throughout the code base. They are part of the `node.Service` interface which, if we'd implement them properly, would be `Register`ed and then they would be called implicitly by the `p2p.Server` boot procedure.

Instead, we call them manually in `swarm.go` and start/stop services manually. I think that is wrong and it should be refactored, @nolash has made similar comments.

In that view, they would be part of the framework, and I think we should not call them manually. That is why I think having a `Close` is justified, which is called via `Stop`, which, if we run unit tests, we should not call as it is conceptually not part of the stack. Instead, we would call `Close` from a test instead.

Reply:
I still don't understand why you want to have this extra layer of `Close`. If you're calling `Close` you can also call `Stop`.

> Instead, we call them manually in swarm.go and start/stop services manually. I think that is wrong and it should be refactored, @nolash has made similar comments.

This unfortunately IMHO is not a refactoring candidate and cannot and should not be automagically started and shut down by the p2p layer _in order to ensure a correct startup/shutdown sequence_. i.e. you might want to shutdown protocols first, before shutting down netstore/localstore. I wouldn't hand over this sequence to the p2p layer as one cannot have any guarantees there. If you insist having it this way then have it your way.

Reply:
I very much agree with what @holisticode is pointing out here. The Swarm design is in general way too monolithic. But to properly execute separating these services is a larger job that has to be well thought through.

> you might want to shutdown protocols first, before shutting down netstore/localstore.

So the stores should be a service aswell, anything in protocols that cannot store and should before ack'ing that it's stored properly, will fail. That sounds perfectly reasonable to me.

Reply:
> my intent was to remove Close, and have the logic sit in Stop. 

Use `close` called by `Stop`